l10n_nl_tax_statement/models/account_move_line.py

```diff
@@ -31,10 +31,55 @@ def _get_l10n_nl_vat_statement_protected_fields(self):
             "tax_tag_ids",
         ]
 
+    def check_field_is_equal(self, changed_protected_field, values):
+        self.ensure_one()
+        field = self._fields.get(changed_protected_field)
+        old_value = self[changed_protected_field]
+        new_value = values[changed_protected_field]
+        if field.type in ["many2many", "one2many"]:
+            # if field is X2M , the only acceptable value is
+            # [[6,0,self[changed_protected_field].ids]]
+            # wich is what the web client posts in case there is a editable X2M in form
+            # that is unchanged.
+            # If it is a list of tuple-commands, and the last one is precisely
+            # values[changed_protected_field][2] == self[changed_protected_field].ids
+            # we will not accept any modification to protected fields,
+            # only the standard 6,0,[ids] coming from the default web edit.
+            return (
+                len(values[changed_protected_field]) == 1
+                and values[changed_protected_field][0][0] == 6
+                and values[changed_protected_field][0][2]
+                == self[changed_protected_field].ids
+            )
+        if new_value:
+            return old_value == new_value
+        return bool(old_value) == bool(new_value)
+
     def write(self, values):
-        if self._l10n_nl_vat_statement_should_check_write(values):
-            self._l10n_nl_vat_statement_check_state()
-        return super().write(values)
+        # before doing anything we check the nl_vat_statement:check_state,
+        if not self._l10n_nl_vat_statement_should_check_write(values):
+            return super().write(values)
+        # now we add code to check if any of the modified fields present in
+        # values are the same as existing field value. this is a known limitation
+        # of odoo web client, it passes all non-readonly value fields in a form
+        # for writing , even if the value has not been changed.
+        protected_fields = self._get_l10n_nl_vat_statement_protected_fields()
+        protected_fields_in_values = [x for x in values.keys() if x in protected_fields]
+        invalid_fields = set()
+        for this in self:
+            for protected_field in protected_fields_in_values:
+                is_equal = this.check_field_is_equal(protected_field, values)
+                if not is_equal:
+                    # if the field is invalid in even one
+                    # of the records it cannot be popped
+                    invalid_fields.add(protected_field)
+        for protected_field in protected_fields_in_values:
+            if protected_field not in invalid_fields:
+                values.pop(protected_field)
+        for this in self:
```

Comment:
One more optimization I just see is to switch the conditions, as _l10n_nl_vat_statement_should_check_write(values) has an @api.model decorator, it is the same for all records:
` ` ` 
if self._l10n_nl_vat_statement_should_check_write(values):
    for this in self:
        this._l10n_nl_vat_statement_check_state()
```

Reply:
An ever more drastic performance enhancement is also possible, but that would involve an even more drastic restructuring of the code:
* Take a subset of the records in self, that do have l10n_nl_vat_statement_id set AND are not in draft state
* Check only for those records whether a protected field has really changed, in that case immediately raise an error (no more needs for an invalid_fields set)
* If an error is not raised, just call super() with the original self

Reply:
true , check_write  is @model, because it conly depends on what keys are present in the write dict. We can keep it out of loop.

First proposal:  we can eliminate that condition all together. We already check that :
```
 59         if not self._l10n_nl_vat_statement_should_check_write(values):                  
 60             return super().write(values)                          
``` 

So i eliminated it altogether:

```
 78         for this in self:                                                               
 79             this._l10n_nl_vat_statement_check_state()                                   
 80         return super(AccountMoveLine, self).write(values)      
```

I  implemented the first  proposal, not the more deep one for this reason-
      
the second restructure, while faster, because it works only on a subset of non-draft state records, does move the logic of _l10n_nl_vat_statement_check_state   outside ,  and put it in this write.
I wrote this with the preference of keeping the logic in the check_state  function...
What we would save is not exponential , but linear.  the check_state is really just a comparison operation.

Reply:
@gfcapalbo You cannot eliminate it altogether. The first check is based on all values passed originally, the second check - that you now eliminated - is based only on the valiues that are changed in one or more records. By eliminating the second check, you undo all the work of checking whether values really changed.

Reply:
@NL66278  Fixed:  https://github.com/OCA/l10n-netherlands/pull/405/commits/2ec172d558c77dc9693e84a6ba2ece7cc34fb086